tensorflow/contrib/go/tensor.go

```diff
@@ -0,0 +1,753 @@
+
+	return t.DoubleVal, nil
+}
+
+// Int32s returns the Tensor content as int32 slice if the tensor
+// type matches, if not returns a ErrInvalidTensorType error.
+// The datatypes are:
+//  - DTUint8
+//  - DTInt8
+//  - DTInt16
+//  - DTInt32
+func (t *Tensor) Int32s() ([]int32, error) {
+	if t.IntVal != nil {
+		return t.IntVal, nil
+	}
+
+	data := t.Data()
+	switch t.DataType() {
+	case DTInt8, DTUint8:
+		t.IntVal = make([]int32, len(data))
+		for i, v := range data {
+			t.IntVal[i] = int32(v)
+		}
+	case DTInt16:
+		numElems := len(data) / cBytesUint16
+		t.IntVal = make([]int32, numElems)
+		for i := 0; i < numElems; i++ {
+			t.IntVal[i] = int32(binary.LittleEndian.Uint16(data[i*cBytesUint16 : (i+1)*cBytesUint16]))
+		}
+	case DTInt32:
+		numElems := len(data) / cBytesInt32
+		t.IntVal = make([]int32, numElems)
+		for i := 0; i < numElems; i++ {
+			t.IntVal[i] = int32(binary.LittleEndian.Uint32(data[i*cBytesInt32 : (i+1)*cBytesInt32]))
+		}
+	default:
+		return nil, &ErrInvalidTensorType{
+			TensorType:   t.DataType(),
+			ExpectedType: DTInt32,
+		}
+	}
+
+	t.Dtype = pb.DataType(TF_TensorType(t.tensor))
+
+	return t.IntVal, nil
+}
+
+// Int64s returns the Tensor content as int64 slice if the tensor
+// type matches, if not returns a ErrInvalidTensorType error.
+// The datatypes are:
+//  - DTInt64
+func (t *Tensor) Int64s() ([]int64, error) {
+	if DTInt64 != t.DataType() {
+		return nil, &ErrInvalidTensorType{
+			TensorType:   t.DataType(),
+			ExpectedType: DTInt64,
+		}
+	}
+
+	if t.Int64Val != nil {
+		return t.Int64Val, nil
+	}
+
+	data := t.Data()
+	numElems := len(data) / cBytesInt64
+	t.Int64Val = make([]int64, numElems)
+	for i := 0; i < numElems; i++ {
+		t.Int64Val[i] = int64(binary.LittleEndian.Uint64(data[i*cBytesInt64 : (i+1)*cBytesInt64]))
+	}
+	t.Dtype = pb.DataType(TF_TensorType(t.tensor))
+
+	return t.Int64Val, nil
+}
+
+// Bools returns the Tensor content as boolean slice if the tensor
+// type matches, if not returns a ErrInvalidTensorType error.
+// The datatypes are:
+//  - DTBool
+func (t *Tensor) Bools() ([]bool, error) {
+	if DTBool != t.DataType() {
+		return nil, &ErrInvalidTensorType{
+			TensorType:   t.DataType(),
+			ExpectedType: DTBool,
+		}
+	}
+
+	if t.BoolVal != nil {
+		return t.BoolVal, nil
+	}
+
+	data := t.Data()
+	t.BoolVal = make([]bool, len(data))
+	for i, v := range data {
+		t.BoolVal[i] = (v == 1)
+	}
+	t.Dtype = pb.DataType(TF_TensorType(t.tensor))
+
+	return t.BoolVal, nil
+}
+
+// GetVal returns the value of the element contained in the specified position
+// in the tensor, Ex: GetVal(1, 2, 3) is equivalent to data[1][2][3] on a
+// multidimensional array.
+// This method returns an error if the number of dimensions is incorrect or
+// are out of range.
+func (t *Tensor) GetVal(i ...int) (val interface{}, err error) {
+	if len(i) != t.NumDims() {
+		return nil, &ErrDimsOutOfTensorRange{
+			TensorDim: t.NumDims(),
+			SpecDims:  len(i),
+		}
+	}
+
+	pos := 0
+	if t.dimWeights != nil {
```

Comment:
This function seems to behave very differently on the first and subsequent calls -- I suggest unifying them.  The error checking is only performed on the first call.  So doesn't this sequence of calls fail?
tensor dims = [2, 2, 2]
t.GetVal(1, 1, 1)  <-- fails at line 482 with ErrIndexOutOfRange
t.GetVal(3, 3, 3)  <-- passes an invalid position to t.getValOnPos,
which does no error checking and will cause a crash.

This would feel a lot more clear to me if the function worked in two steps:

1)  if t.dimWeights != nil {  ... populate the weights, which can be done independently of i }

2)  Calculate and validate the pos using t.dimWeights